### ISSUE TITLE
Marlin 2.0 RC1 - TODO

### ISSUE THREAD

**Issue by thinkyhead** [MarlinFirmware/Marlin#14345]
Let's coordinate the Marlin 2.0 RC1 release and track remaining work for this first release candidate.

### AVR

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[RAMPS](https://reprap.org/wiki/RAMPS_1.4), [RAMBo](https://reprap.org/wiki/Rambo) etc.|AVR|💚|💚|💚|💚|💚|💚|💚|💚

### SAM

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[Due](https://www.arduino.cc/en/Guide/ArduinoDue), [RAMPS FD](https://reprap.org/wiki/RAMPS-FD), etc.|SAM3X8E|💚|💚|💚|💚|💚|💚|💚|💚
[Duet 2 Wifi](https://duet3d.com/DuetWifi) + X5|SAM4E8E|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑
[Duet 2 Maestro](https://www.duet3d.com/Duet2Maestro)|SAM4S8C|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️
[Archim 1.0](https://reprap.org/wiki/Archim_v1.0)|SAM3X8E|💚|💚|💚|💚|💚|💚|💚|💚
[Archim 2.0](https://reprap.org/wiki/Archim2)|SAM3X8E|💚|💚|💚|💚|💚|💚|💚|💚
[Grand Central](https://www.adafruit.com/product/4064?fbclid=IwAR1eOOmapZ09otb1oK9RmFBqItBR3Ndsb7RMxbN0lYH8dx8HlQ2fAdNNfIs)|ATSAMD51|💚|💚|⚠️|⚠️|💚|🛑|🛑|💚

### LPC

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[Re-ARM](https://www.panucatt.com/Re_ARM_for_RAMPS_p/ra1768.htm)|LPC1768|💚|💚|💚|💚|💚|💚|💚|💚
[MKS-SBASE](https://github.com/makerbase-mks/MKS-SBASE)|LPC1768|💚|💚|💚|💚|💚|💚|💚|💚
[SKR v1.3](https://github.com/bigtreetech/BIGTREETECH-SKR-V1.3)|LPC1768|💚|💚|💚|💚|💚|⚠️|⚠️|⚠️
[Smoothieboard](http://smoothieware.org/smoothieboard)|LPC1769|💚|💚|💚|💚|💚|💚|💚|💚
[Azteeg X5 GT](https://www.panucatt.com/azteeg_X5_GT_reprap_3d_printer_controller_p/ax5gt.htm)|LPC1769|💚|💚|💚|💚|💚|💚|💚|💚
[Cohesion3D Remix](https://cohesion3d.com/)|LPC1769|💚|💚|💚|💚|💚|⚠️|⚠️|⚠️
[Selena Compact](https://github.com/Ales2-k/Selena)|LPC1768|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️

### STM32F1

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[Malyan M200](http://malyansys.com/product/m200-v1/)|STM32F103C8|⚠️|⚠️|n/a|n/a|⚠️|⚠️|⚠️|⚠️
[MKS Robin 2.4](https://github.com/makerbase-mks/MKS-Robin)|STM32F103ZET6|💚|💚|🛑|?|💚|🛑|🛑|🛑
[Chitu3D V3.9](https://drive.google.com/drive/folders/0B-iwzoALEv3kbnVHVFZxemZ6TDQ)|STM32F103ZET6|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑
[Longer3D LK1](https://www.aliexpress.com/item/32944556459.html)|STM32F103VE|💚|💚|n/a|n/a|💚|n/a|n/a|n/a
[SKR Mini v1.1](https://github.com/bigtreetech/BIGTREETECH-SKR-MINI-V1.1)|STM32F103RC|💚|⚠️|⚠️|⚠️|💚|n/a|n/a|🛑

### STM32F4

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[STEVAL-3DP001V1](https://www.digikey.com/product-detail/en/stmicroelectronics/STEVAL-3DP001V1/497-16536-ND/6023689)|STM32F401VE|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑
[ARMed]|STM32F401VE|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑

### STM32F7

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[The Borg 3D](https://github.com/Spawn32/The-Borg-3D)|STM32F765ZGT6|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️

### ESP32

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[Espressif 32 Wifi](https://www.espressif.com/en/products/)|ESP32|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️|⚠️

### Teensy

Board|MCU|SD Card|LCD|TMC SPI|TMC UART|EEPROM|WiFi|Max6675|Neopixel
-|-|-|-|-|-|-|-|-|-
[Teensy 3.5](https://www.pjrc.com/store/teensy35.html)|MK64FX|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑
[Teensy 3.6](https://www.pjrc.com/store/teensy36.html)|MK66FX|🛑|🛑|🛑|🛑|🛑|🛑|🛑|🛑

#### Legend

- 💚 (good) Working, no major issues.
- ⚠️ (beta) Working, but not completely.
- 🛑 (alpha) Not yet working.
- ❓ (unknown) You tell me!

### Checklist for RC1:

- Compiles and runs on…
  - [x] AVR: many boards
  - [x] DUE: many boards
  - [x] LPC176x: many boards
  - [ ] ESP32: one or two boards
  - [ ] STM32: Needs software serial
  - [x] STM32F1 - a few boards
  - [x] STM32F4 - a few boards
  - [ ] STM32F7 - one or two boards
  - [x] Teensy 3.1/3.2 - Printrbot, et. al.
  - [ ] Teensy 3.5/3.6 - alpha

- `FAST_PWM_FAN` (e.g., Native PWM, also good for spindle/laser):
  - [x] AVR
  - [x] DUE
  - [ ] LPC176x
  - [ ] ESP32
  - [ ] STM32
  - [ ] STM32F1
  - [ ] STM32F4
  - [ ] STM32F7 - one or two boards
  - [ ] Teensy 3.1/3.2
  - [ ] Teensy 3.5/3.6

**Comment by gloomyandy** [comment]
I'm not sure what the definition is of a "Supported platform", but we may want to add the Bigtreetech SKR V1.3 board (possibly the most popular 32bit board at the moment) and probably some of the other SKR boards (which use various STM processors and are popular, but possibly not as well supported at present). How well these popular 32bit boards work is probably a good indication of the current state of "real world" 32bit support and should be tracked as best we can. Perhaps those boards should also be a major focus for RC1?

**Comment by Filcanna** [comment]
![Schermata 2019-06-21 alle 11 31 58](https://user-images.githubusercontent.com/26361956/59914021-c0b9b500-9419-11e9-86e1-f5d1f7209d36.png)
Pin scheme for MCU on SKR Pro v 1.1,  I think

**Comment by Bob-the-Kuhn** [comment]
The MKS Robin 2.4 (STMF103ZET6) may be a candidate for the list.  It compiles via PlatformIO and runs Marlin.  I don't know how functional it is.

**Comment by hobiseven** [comment]
Alfawise u20/20+/u30 is a perfect candidate! Stm32f1vet6.
Runs perfectly , BL touch and touchmi enabled and tested
Compiles on platformio.
Qvga tft direct 16 bit parallel interface using dogm “zoom”
Resistive touch screen based on standard ads7843. ( driver can be improved but it works with 3 buttons émulation and is software spi controled).
And Hal debugged!!!
Sd card support ok
Fan soft pwm
Flexible BL touch configuration for cpu pins allowing multiple board revisions to be compatible
Serial port 250k through usb. Octoprint ok
Eeprom in the works

In a nutshell : status = excellent!

**Comment by Phr3d13** [comment]
GTM32 variants are getting close. Just got my hands on a Geeetech A30 (GTM32 mini) today. Geeetech Rostock 301 (GTM32_PRO_VB) is really close (I think it's down to one or two issues - eeprom and temps). I'll do some more testing this weekend.

**Comment by Patag** [comment]
Does the coming soon BigtreeTech SKR Pro is so much too far from the current Marlin status ?
I still don't have the "big picture" of Marlin V2, so sorry in advance if it's a silly question. In fact, I even don't know if STM32F407ZGT6 already exists in Marlin ecosystem.

**Comment by Roxy-3D** [comment]
> Please add your suggestions for the things we should require before committing to a first release candidate.

If we think back to when the v2.0.0 branch was created, its purpose was to provide a place to do the work to get Marlin migrated to 32-bit platforms.   And at that time, the original thinking was we needed it solid and working on a '32-bit Reference Platform'.   We were thinking that if we could get it moved successfully to one 32-bit board, others would follow.

Then...   the hierarchical organization work started.   And it just made sense to do that in the v2.0.0 branch because that work needed to be separated out of the normal work flow and usage.   That made sense, but it made the v2.0.0 branch dual purpose.

I'm of the opinion the v2.0.0 is ready to release and is suitable to be used to start the next branch of Marlin's evolution.   I don't have strong feelings on it, but we have multiple 32-bit processors supported and the hierarchical file organization is pretty well proven at this point.

To my way of thinking, all of the "My favorite 32-bit processor is almost ready, lets add it to the list." comments don't have much merit.   We are going to keep a check off list detailing the status of all the various processors anyways.    As a particular processor family's status changes...  we can update the list.

My personal opinion is:   **WE ARE GOOD TO GO.**

**Comment by gloomyandy** [comment]
Hmm perhaps to try and get some sort of focus it would be useful to have a feature list that can be used to rate how complete support on a particular board/processor is. So for me some key features are...

- SD card support
- TMC SPI and UART driver support
- LCD support - Graphical and character, plus perhaps some of the new ones like FYSETC_MINI_12864
- BLTOUCH support
- EEPROM support
- USB serial support
- Support for UART based touch displays (basically a working UART).

I'm assuming that all boards will support things like heaters and basic stepper operation. But I think these days most folks will probably expect to be able to use some/all of the above (assuming the board is capable). I've almost certainly missed some items that should be on the list but I've tried to capture the ones that I know can be tricky to get working.

I have no real idea how well the various HALs stack up against the above list. From my own experience I'd say that the LPC176x based boards support pretty much everything above, but I've no idea about the STM based boards. Such a list might also be of interest to potential contributors giving them some idea of what areas need work. Though I guess most features tend to get added by someone that feels a need for it and that has the hardware.

**Comment by Roxy-3D** [comment]
I think I agree this would be a very good chart to have on the  README.md for Marlin v2.0.0.  For each processor, detail the status of:

- SD card support
- TMC SPI and UART driver support
- LCD support - Graphical and character, plus perhaps some of the new ones like FYSETC_MINI_12864
- BLTOUCH support
- EEPROM support
- USB serial support

It might also make sense to add information about the point release where that support became solid and trustworthy (on a per line item basis).

**Comment by InsanityAutomation** [comment]
> 
> 
> Does the coming soon BigtreeTech SKR Pro is so much too far from the current Marlin status ?
> I still don't have the "big picture" of Marlin V2, so sorry in advance if it's a silly question. In fact, I even don't know if STM32F407ZGT6 already exists in Marlin ecosystem.

It exists. I have a board on the way. If its just a matter of a pin file, config samples ect and the HAL is good, it just might make it in. Time will tell.

**Comment by InsanityAutomation** [comment]
> My personal opinion is: **WE ARE GOOD TO GO.**

The only blocker I see is still getting layer shifts on machines with 2.0 that reverting to 1.1.9 with a matching as possible config resolves. If this got sorted out id say just pick at a bunch of small things for a clean snapshot then go. Ill browse the issue queue for anything else that jumps out this weekend as a potential blocker and post the number here if I come across any.

**Comment by Roxy-3D** [comment]
>> My personal opinion is: WE ARE GOOD TO GO.

> The only blocker I see is still getting layer shifts on machines with 2.0 that reverting to 1.1.9 with a matching as possible config resolves. If this got sorted out id say just pick at a bunch of small things for a clean snapshot then go.

Actually...  I agree with this thinking.  But my perspective is a little bit different.   At a higher level, if we are going to be adding things to a list of things that need to be resolved prior to releasing v2.0.0....

Resolving the Layer Shift issue is at the top of the list.

**Comment by gloomyandy** [comment]
This is probably off topic, but I'm totally confused about the layer shift issue. Do we actually have any hard examples now (I know we had some that seemed to be down to an STM timer issue - hopefully resolved, when that PR is merged)? So many of them seemed to turn out to be down to users switching to different stepper drivers or making other changes, or just mechanical issues. I notice that the thread has been closed.

**Comment by Phr3d13** [comment]
I agree with the general consensus, list the boards/machines that work 100% and maybe list what needs work on what boards/machines still. But I think you should make a 2.0 release.

**Comment by Roxy-3D** [comment]
> This is probably off topic, but I'm totally confused about the layer shift issue. Do we actually have any hard examples now (I know we had some that seemed to be down to an STM timer issue - hopefully resolved, when that PR is merged)? 

Nope!   Definitely not off topic.   If we are talking about action items that need to be completed prior to doing the v2.0.0 release, this is  'On Topic'.  

I think the layer shifts are real.   Most of them are caused by various mechanical issues with the printer.   Some are caused because people are pushing their feed, acceleration and jerk numbers too high.  (And just because it 'seems to work', that doesn't mean with a worst case GCode file the printer can actually handle it.)

But here is the thing...   The amount of layer shift issues we are aware of on the vetted 32-bit platforms is almost zero.    The bulk of the layer shift problems we are seeing are on AVR processors.   And I know I can avoid layer shifts if I scale my F/R settings on the LCD Panel down to 50%.    Above 50%, I will some times see a layer shift on a long print.

That sort of points to the AVR processors not having enough muscle to get everything done (at interrupt time) and some how we are losing steps.    Right now, the people focused on the problem are not making good headway in resolving the root cause.

**Comment by hobiseven** [comment]
Well 
Layer shift is solved on stm32f1...

**Comment by Ghenghis** [comment]
please add support for Bigtreetech SKR v1.3 & Pro v1.1 boards.

**Comment by tpruvot** [comment]
well "fixed" with the unmerged PR #14030 but our touchscreen is not in Marlin neither

**Comment by Grogyan** [comment]
I'm still having trouble with Max31855 support, on both ReArm and SKR 1.3.

However, since getting the SKR 1.3 I'll be using this as my reference now

**Comment by InsanityAutomation** [comment]
> 
> 
> This is probably off topic, but I'm totally confused about the layer shift issue. Do we actually have any hard examples now (I know we had some that seemed to be down to an STM timer issue - hopefully resolved, when that PR is merged)? So many of them seemed to turn out to be down to users switching to different stepper drivers or making other changes, or just mechanical issues. I notice that the thread has been closed.

One big indicator is on IDEX machines when we see both X axis heads shift together. Non-firmware issues cant do that. Most likely thats a planner or ISR output issue. These are the scenarios we can be absolutely certain are software and is a solid indicator. We have several examples of prints with a large number of short jerky moves that shift every time. To date every user who has reverted to 1.1.9 has seen these shifts cease immediately. The sheer number of reports from dealing with a particular manufacturer who is shipping machines with 2.0 preinstalled and the distributor for them in my region has been fairly staggering.

**Comment by InsanityAutomation** [comment]
> 
> 
> please add support for Bigtreetech SKR v1.3 & Pro v1.1 boards.

1.3 is already pretty well supported. See my statement above on the 1.1 boards. Can get an answer sooner if you can get btt hope to ship mine faster!

**Comment by tpruvot** [comment]
Could we reserve some "ID" for the longer3D/alfawise STM32F1 board ? to avoid to change it each month ? :p

**Comment by Roxy-3D** [comment]
> One big indicator is on IDEX machines when we see both X axis heads shift together. Non-firmware issues cant do that. Most likely thats a planner or ISR output issue. These are the scenarios we can be absolutely certain are software and is a solid indicator. 

Agreed!   Marlin v2.0.0 has a layer shift bug that is not that hard to duplicate on 8-bit AVR processors.   

Should we add its resolution to the check off list for releasing v2.0.0 ?

**Comment by boelle** [comment]
> > One big indicator is on IDEX machines when we see both X axis heads shift together. Non-firmware issues cant do that. Most likely thats a planner or ISR output issue. These are the scenarios we can be absolutely certain are software and is a solid indicator.
> 
> Agreed! Marlin v2.0.0 has a layer shift bug that is not that hard to duplicate on 8-bit AVR processors.
> 
> Should we add its resolution to the check off list for releasing v2.0.0 ?

seems wise to do, you get my vote

and maybe go through the issue list and pick confirmed issues that are serious enough and add them to the list also

**Comment by boelle** [comment]
i think this one is a good one to look at https://github.com/MarlinFirmware/Marlin/issues/4931

check config at boot and either warn or limit speed so it matches with what the cpu is able to

and this one since it's on the 2.0.0 milestone https://github.com/MarlinFirmware/Marlin/issues/5079.

Even thou this is a feature request https://github.com/MarlinFirmware/Marlin/issues/6199 i think it should be looked at anyway since its so a common bed these days (prusa is not the only one anymore)

and this one has the deisgn concept label and is still open even thou its hinted  that its not likely to get implemented, maybe do a quick look at it? https://github.com/MarlinFirmware/Marlin/issues/6295

also with design concept https://github.com/MarlinFirmware/Marlin/issues/8195

@Roxy-3D came with a good suggestion here https://github.com/MarlinFirmware/Marlin/issues/7274 but it has not been confirmed yet

these might also be relavant: https://github.com/MarlinFirmware/Marlin/issues/8497
https://github.com/MarlinFirmware/Marlin/issues/9048
https://github.com/MarlinFirmware/Marlin/issues/9205
https://github.com/MarlinFirmware/Marlin/issues/9403
https://github.com/MarlinFirmware/Marlin/issues/9742
https://github.com/MarlinFirmware/Marlin/issues/9917
https://github.com/MarlinFirmware/Marlin/issues/9931
https://github.com/MarlinFirmware/Marlin/issues/10010
https://github.com/MarlinFirmware/Marlin/issues/10455
https://github.com/MarlinFirmware/Marlin/issues/10459
https://github.com/MarlinFirmware/Marlin/issues/10549
https://github.com/MarlinFirmware/Marlin/issues/10707
https://github.com/MarlinFirmware/Marlin/issues/11262
https://github.com/MarlinFirmware/Marlin/issues/11263
https://github.com/MarlinFirmware/Marlin/issues/13130
https://github.com/MarlinFirmware/Marlin/issues/13201

**Comment by Bergerac56** [comment]
And we have still the linear advance issue (at least for non TMC drivers) #11205

**Comment by thisiskeithb** [comment]
~Hopefully we can add the [MKS Robin Nano](https://github.com/MarlinFirmware/Marlin/pull/14363) & [MKS Robin Mini](https://github.com/MarlinFirmware/Marlin/pull/14366) (both STM32F103VET6) to the supported boards list.~

Boards are now supported! :smiley:

**Comment by AnHardt** [comment]
> SD card support
TMC SPI and UART driver support
LCD support - Graphical and character, plus perhaps some of the new ones like FYSETC_MINI_12864
BLTOUCH support
EEPROM support
USB serial support

Is looking very implementation/feature centric to me. I'd prefer a more functionality centric view.

To make work a FDM 3D printer (I hope thast's still Marlins main target.) the software (Marlin) and the board (HAL), together, have to reliably support at least.
* smooth movement on >= 4 steppers / stepper drivers (regardless of the drivers type, It only needs one type to work) 
* \>= 1 temperature sensor (regardless of the sensor type)
* \>= 1 heater (regardless of the PWM type)
* \>= 1 heater algorithm (not necessary PID)
* the ability to home  3 cartesian axes
* configuration storage (regardless of if implemented in EEPROM, battery buffered SRAM, PROM, internal or external SD-card, USB-stick, ... ) 
* ...

and eiter 
* \>= 1 (serial) host interface (could be: TTL serial, RS232 serial, USB, Bluetooth, Ethernet, WiFi, ...)

and or
* * \>= 1 local mass storage device (SD-card, USB stick, harddisk, on board SPI flash chip, ...)
* * with a display and input device (display board) to select a file to print 

The minimum required security functionalitys are:
* Watchdog timer to check is the (heater) system is still running
* Heater system reliability check (Heater protection against bad electrical/thermal contact sensors/heaters)
* Power supply OFF functionality at least for the heaters power supply to have an alternate way to switch off the heaters.

and less important if the motors are not powerful enough to destroy the machine
* hardware security endstops (ideally on both ends).
* ...

With this functionality, a hobbyist (and off cause a professional system integrator)  can build a limited but functional printer.
So this could be the minimum **required functionality** to call a board/system/... **supported by Marlin 2**.  

Everything additional is a comfort feature and is not strictly required.
It may be helpful for the users to keep a **list of additionally working features** per platform (or whatever makes sense).

**Comment by xC0000005** [comment]
Do we require a board to build in PlatformIO to be functional? 
The M200 V1 and V2 are basically fully functioning - these boards will never support Neopixel & other fun hardware addons - they're extremely limited, with fixed drivers, etc.
All that's really missing is a few minor LCD fixes to deal with a new timeout feature in the delta and V3. 
Also, the V1 and V2/V3 run on different processors - V1 is an STM32F103, V2/Pro (and the delta) run on STM32F070.

**Comment by Hedda** [comment]
BIQU BIGTREETECH SKR MINI-E3 board uses STM32F103RCT6 with TMC2209 stepper drivers:

https://www.reddit.com/r/BIGTREETECH/comments/c0hdht/bigtreetech_skr_mini_e3_done/

FYSETC Cheetah board uses STM32F103 (of an unknown model) with TMC2209 stepper drivers:

https://www.reddit.com/r/3Dprinting/comments/c1lttc/fysetc_cheetah_board_chinaclone_of_th3d_ezboard/

Does anyone know what 32-bit MCU that TH3D Studio's EZBoard Lite use? Said to run at 120Mhz:

https://www.reddit.com/r/3Dprinting/comments/bvsd1q/ezboard_lite_formerly_tough_controller_for/

**Comment by InsanityAutomation** [comment]
@Hedda Its an LPC1768 on the TH3D board.

**Comment by Hedda** [comment]
Does not the Duet 2 Maestro use the Atmel SAM4S8C (ATSAM4S8C) MCU and not the SAM3X8E? Typo?

https://duet3d.dozuki.com/Wiki/Duet_2_Maestro_Hardware_Overview

And the Duet 2 WiFi and Duet 2 Ethernet should be using similar use the Atmel SAM4E8E (ATSAM4E8E)?

https://duet3d.dozuki.com/Wiki/Hardware_Overview

**Comment by trouch** [comment]
Can't wait for RC, keep up with your great work everyone ! 
I have a Fysetc Cheetah on its way, will keep posted.

**Comment by Hedda** [comment]
@InsanityAutomation if it runs at 120 MHz then more likely an LPC1769 model then an LPC1768 model

https://www.nxp.com/docs/en/data-sheet/LPC1769_68_67_66_65_64_63.pdf

LPC1769 are the only models in that series of NXP MCU to officially run up to 120 MHz.

**Comment by Hedda** [comment]
Support for FYSETC CHEETAH board committed FYSETC's own Marlin 2.0 fork on GitHub 4 days ago:

https://github.com/FYSETC/Marlin-2.0.x-FYSETC/commit/4fa345a762717daf81e794f2594978cf052248c1

**Comment by simon-jouet** [comment]
For the ESP32 there is no commercial board as of now (at least none that i'm aware of). The board that is currently the most "common" is the R1 and the R2 is being worked on:

- https://github.com/simon-jouet/ESP32Controller R1 is very basic, it's the board that @thinkyhead, @p3p, @grownseed and have a few more people
- https://github.com/simon-jouet/ESP32Controller/tree/r2 Is the new design that uses I2S to have more output pins (and improve the stepping), I haven't made any PCB for it yet (working on it) - feel free to comment on the open issues if you are interested

- @vivian-ng has made https://github.com/vivian-ng/Hardware/issues/2 based on my R1 version (no I2s yet)
- @felixstorm made a board based on the R2 https://github.com/felixstorm/Creality_Ender_3_ESP32_Board
- @ringle  made a PCB for the R2 https://github.com/Ringel/ESP32Controller/tree/r2

**Comment by felixstorm** [comment]
From my personal experience, current Marlin ESP32 / I2S code seems to be working well (with my [ESP32 board](https://github.com/felixstorm/Creality_Ender_3_ESP32_Board) based on @simon-jouet's R2 and my stock Ender 3), but since I currently do not have much spare time, I only made a few prints yet.

Personal experience so far:

|Feature|Status|
|---|---|
|SD card|working fine|
|TMC SPI and UART driver|not in use or tested|
|LCD|working fine (Ender 3 stock display)|
|BLTOUCH|working fine|
|EEPROM|working fine|
|USB serial|working fine|
|Fans|working fine (hotend, part cooling and controller controlled separately)|
|Heater|working fine (1 in use)|
|Heated Bed|working fine|
|WiFi|not really used yet|

Issues I've had so far:  
* Bed thermal protection kicked in once. Did not have the time to do research on this, so I increased period and hysteresis a bit and did not have any further occurrences so far.
* Layer shift: I had one single layer shift with the very first (and very large) print that I ran overnight, which I did never have before with the stock controller and Marlin 1.1.9. Again I could not do any research, but it also did not happen again.

*So my personal opinion: "beta" sounds fine to me for ESP32 support.*  
Maybe someone else (e.g. @grownseed, @Idorobots) can shed some light on WiFi status?

**Comment by trouch** [comment]
> Support for FYSETC CHEETAH board committed FYSETC's own Marlin 2.0 fork on GitHub 4 days ago:
> 
> [FYSETC/Marlin-2.0.x-FYSETC@4fa345a](https://github.com/FYSETC/Marlin-2.0.x-FYSETC/commit/4fa345a762717daf81e794f2594978cf052248c1)

I'm looking to merge that into a bugfix-2.0.x branch before submitting a PR. Lot of changes and several ones conflicting with recent support of TMC2209 here. They also added SD support and other tweaks. GIT diff returns 12 conflicting files, but I can see runtime conflicts as well by digging into changes.

**Comment by InsanityAutomation** [comment]
@Hedda yup, typo sorry it is a 1769

**Comment by Idorobots** [comment]
WiFi is usable as an additional serial right now (err, more like two months back, as I haven't had much time to invest into ESP32 port since about then). The UI allows you to run G-code commands, but there's the usual multiple-serial shenanigans going on - some commands print to one of them, some others to both, etc. It's entirely usable without the WiFi.

**Comment by vivian-ng** [comment]
Regarding WiFi for ESP32, I have been using @luc-github Marlin [fork](https://github.com/luc-github/Marlin) (which is basically Marlin with his ESP3D webUI) in the testing of my board. It works fine for the few test prints I have done, but I have yet to do any intensive testing (in terms of long-hour prints) yet. I think @luc-github has recently [filed a FR](https://github.com/MarlinFirmware/Marlin/issues/14390) here about how to merge his work back into Marlin. Coupled with the great work @simon-jouet has done in implementing I2S, and @felixstorm for fixing issues with the HAL, my prototype board can now both use an LCD controller and also a webUI.

**Comment by ktand** [comment]
The STM32F407 based board [ARMED](https://github.com/ktand/Armed) is fully working using the HAL_STM32:

SD card support
TMC SPI driver support
LCD support - Graphical and character, (FYSETC_MINI_12864 not yet working)
EEPROM support
USB serial support

**Comment by OneOfEleven** [comment]
> 
> 
> The STM32F407 based board [ARMED](https://github.com/ktand/Armed) is fully working using the HAL_STM32:
> 
> SD card support
> TMC SPI driver support
> LCD support - Graphical and character, (FYSETC_MINI_12864 not yet working)
> EEPROM support
> USB serial support

Is there no TMC serial driver (for all the uart TMC'ers out there) ?

**Comment by Msq001** [comment]
> > The STM32F407 based board [ARMED](https://github.com/ktand/Armed) is fully working using the HAL_STM32:
> > SD card support
> > TMC SPI driver support
> > LCD support - Graphical and character, (FYSETC_MINI_12864 not yet working)
> > EEPROM support
> > USB serial support
> 
> Is there no TMC serial driver (for all the uart TMC'ers out there) ?

yeah,  no software serial in stm32 now,  but it will be added later

**Comment by Gamester17** [comment]
Anyone working on a 3D-printer controller board based on the new STM32MP1 MPU series?
https://blog.st.com/stm32mp1-mpu-stm32mp157a-ev1-stm32mp157c-dk2/

STMicroelectronics claims its architecture enables developers to use the same software as STM32 MCU series.

For technical specification summery checkout:
https://www.cnx-software.com/2019/02/21/stmicro-stm32mp1-cortex-a7-m4-mpu/

STM32MP157A-DK1 and STM32MP157C-DK2 Discovery kits are only $99 with LCD and $69 without
https://www.st.com/en/evaluation-tools/stm32mp157a-dk1.html
https://www.st.com/en/evaluation-tools/stm32mp157c-dk2.html
STM32MP157A-EV1 Evaluation kit for PCB engineers does however have a much steeper price at $399
https://www.st.com/en/evaluation-tools/stm32mp157a-ev1.html

First read about this MCU/MPU in the release news about PanGu board from I2Som
https://www.cnx-software.com/2019/06/25/pangu-board-stm32mp1-sbc/

**Comment by oneofnine** [comment]
> Anyone working on a 3D-printer controller board based on the new STM32MP1 MPU series?
> https://blog.st.com/stm32mp1-mpu-stm32mp157a-ev1-stm32mp157c-dk2/

You could forget marlin and run octo print on the two A7 cores with klipper on the M4 core :)

**Comment by thinkyhead** [comment]
> the usual multiple-serial shenanigans going on - some commands print to one of them, some others to both

It would be good to find a savvy developer who has a setup with wifi (or other multiple serial port hosts) and get down to debugging the serial redirection. Command responses are supposed to be directed out to the serial port that sent the command. General machine state output, not connected to a particular command, should go to both serial ports. Serial port redirecting is meant to always be temporary and automatically restored when a block that changed it goes out of scope.

A thorough day or three of attentive testing and debugging of serial port redirection is needed to get that output into shape.

**Comment by luc-github** [comment]
>the usual multiple-serial shenanigans going on - some commands print to one of them, some others to both

Is that specific to ESP32 ? 
If yes it may linked to Websocket instead. And likely due to usage of AsyncWebSocket which reject message if it cannot proceed it (https://github.com/me-no-dev/ESPAsyncWebServer/pull/411)
I do not have such issue on my side since I use another websocket library

**Comment by thinkyhead** [comment]
> We have several examples of prints with a large number of short jerky moves that shift every time.

This is excellent! If there's a G-code file that reproduces the issue at the same point every time, then we can just add a bit of debugging to see exactly what's going on. The race condition from <=1.1.8 was too random to debug this way. @InsanityAutomation — Can you send me the G-code to test?

**Comment by xC0000005** [comment]
> > the usual multiple-serial shenanigans going on - some commands print to one of them, some others to both
> 
> Is that specific to ESP32 ?

I have a fork which enables the Malyan M200 WIFI printing (no surprise, it's the ESP8266's Telnet/Serial bridge, configured as a second serial port), and I see the behavior there. Since that doesn't use the websockets code, I suspect it's more likely to do with the redirection code, though the Rememberer template sure looks like it ought to work.

**Comment by thinkyhead** [comment]
I've updated the OP with some more tables. I could go down to the every-board level, but I'm lazy!

**Comment by DavidThijs** [comment]
Hi, Bigtreetech just released the SKR1.1 PRO which is using a STM32F4 Cortex M4 CPU.
This looks promising. Will Marlin 2.0 RC-1 support that new mainboard ?

**Comment by InsanityAutomation** [comment]
I have one on the way, so we'll see when it gets here.

**Comment by weed2all** [comment]
![Screenshot_20190709-082020_Samsung Internet](https://user-images.githubusercontent.com/33079342/60867485-dbb26480-a222-11e9-9476-346d3ca5693d.jpg)
So I get back to this as is unclear for me...marlin 2.0 rc1 states that have fully work tmc uart on due board!so I still have my radds and due board!can I use 4 tmc2208 drivers in uart with it?

**Comment by jmz52** [comment]
[ARMed Stick](https://hackaday.io/project/160173-armed-stick-stm32-arm-cortex-m3-development-board) is '`powered by the popular STM32F103C8T6 micro controller`', so it is not STM32F4 but an STM32F1

**Comment by thinkyhead** [comment]
> the Rememberer template sure looks like it ought to work

Tracking down the exact spot where it sets the port wrong should be an interesting challenge.

**Comment by weed2all** [comment]
So what can I do to test it?I have the drivers set for uart...the radds is up to date with the latest marlin 2.0.x branch...

**Comment by thinkyhead** [comment]
> ARMed Stick is 'powered by the popular STM32F103C8T6 micro controller', so it is not STM32F4 but an STM32F1

We'll have to ask the author of #12147 @ktand about that.

**Comment by weed2all** [comment]
Now I have all my printers(almost all)with duet board, but I want to have my dolly printer where i can test boards and firmwares...

**Comment by thinkyhead** [comment]
> So what can I do to test it?
- https://www.youtube.com/watch?v=GtlsW3FDN3E
- ```diff
  #if NUM_SERIAL > 1
    extern int8_t serial_port_index;
    #define _PORT_REDIRECT(n,p)   REMEMBER(n,serial_port_index,p)
    #define _PORT_RESTORE(n)      RESTORE(n)
    #define SERIAL_OUT(WHAT, V...) do{ \
      if (!serial_port_index || serial_port_index == SERIAL_BOTH) (void)MYSERIAL0.WHAT(V); \
      if ( serial_port_index) (void)MYSERIAL1.WHAT(V); \
    }while(0)
  + #define SERIAL_ASSERT(P)      if(serial_port_index!=(P)){ debugger(); }
  #else
    #define _PORT_REDIRECT(n,p)   NOOP
    #define _PORT_RESTORE(n)      NOOP
    #define SERIAL_OUT(WHAT, V...) (void)MYSERIAL0.WHAT(V)
  + #define SERIAL_ASSERT(P)      NOOP
  #endif
  ```

Place `SERIAL_ASSERT(0)` where the serial port should be "0" and `SERIAL_ASSERT(1)` where the serial port should be "1" and so on. Replace `debugger()` with some function (or macro) that gives an indication where the error happened.

The debugger is the most direct way to see what's going on. But you could have the `SERIAL_ASSERT` macro print out to the serial port something fun like `PSTR(__FILE__ " : " __LINE__)` which the compiler fills in with the right strings.

**Comment by jmz52** [comment]
> We'll have to ask the author of #12147 @ktand about that.

Oh.
There is an [ARMed STICK](https://hackaday.io/project/160173-armed-stick-stm32-arm-cortex-m3-development-board) based in STM32F103C8 controller and there is an [Arm'ed 3D Printer Controller](https://github.com/ktand/Armed) based on STM32F407VE controller.

Link is the original post got me confused as it is pointing to ARMed STICK i.e. STM32F1.

**Comment by ktand** [comment]
> > ARMed Stick is 'powered by the popular STM32F103C8T6 micro controller', so it is not STM32F4 but an STM32F1
> 
> We'll have to ask the author of #12147 @ktand about that.

I'm the designer of the ARMED board (https://github.com/ktand/armed). Defined in Marlin's boards.h as:

#define BOARD_ARMED                   4104

and is based on the STM32F407 controller.

**Comment by ktand** [comment]
@thinkyhead, as mentioned before, the following features are stable for the ARMED board:

- SD Card
- LCD
- TMC SPI
- EEPROM

The board uses a STM32F407VET6 controller, not a STM32F401VE.

Board link:
[Armed 3D Printer Controller](https://github.com/ktand/Armed)

**Comment by GMagician** [comment]
Lol, Adafruit grand central internal eeprom is working, onboard sd card is on the road, LCD (at least on ruramps can't use sd detect pin because is used by SPI itself. A cable patch is needed) but I think I'll not be able to test nothing more nor TMC in SPI/USART nor neopixel, nor max6675 (what is it?), wifi (?!? on board support or some external board? ruramps has a esp8266 connection but is this required?)

Also external eeprom on shield should works but I'm expecting a ruramps and I don't know when it will be sent.

With these limits, Is there any possibility it will be added to Marlin repo?

**Comment by Roxy-3D** [comment]
> With these limits, Is there any possibility it will be added to Marlin repo?

Don't worry about it being supported in the first release.   First, it probably is going to take several more months to get things clean enough for that to happen.   But if it doesn't make the first release....   It will be included as soon as it is ready.    You aren't going to be left behind.

**Comment by xC0000005** [comment]
We need an icon for "not applicable." 
On the Malyan M200 V1/V2/V3/Pro and M300 printers, there will never be TMC in either flavor (these are fixed driver boards). There's no free SPI output, so no Max support. on V1s there's not even a free GPIO to hook a Neopixel to once the Arduino Core and library support it.

Summary of different Malyan M200/M300 printers:
Malyan M200 V1 - SMT32F103
LCD has minor bugs.
EEPROM, SD, work.
Max, TMC, Wifi, Neopixel aren't possible.

Malyan M200 V2 - SMT32F070
LCD has minor bugs.
EEPROM, SD, work.
Max, TMC, Wifi aren't possible.
Neopixel needs work (lots and lots of work).

The V3 is essentialy the V2 + an auto-level sensor, in the same state (same MCU). I'd list it as V2/V3.

**Comment by MS1987** [comment]
**Description**
Initial MKS Robin Lite and MKS Robin Lite2 (STM32F103RCT6) support brought over from Makerbase's [Robin Lite repo](https://github.com/makerbase-mks/MKS-Robin/tree/master/MKS%20Robin%20Lite) and [Robin Lite2 repo](https://github.com/makerbase-mks/MKS-Robin/tree/master/MKS%20Robin%20Lite2)

**Benefits**
Brings MKS Robin Lite/Lite2 board support into Marlin.

**Related Issues**
None.

**Comment by berezovskyi-oleksandr** [comment]
@thinkyhead first message mentions that FAST_PWM_FAN works on Due. 

At `Marlin/src/HAL/HAL_DUE/SanityCheck.h` there is condition preventing build:
```
#if ENABLED(FAST_PWM_FAN)
  #error "FAST_PWM_FAN is not yet implemented for this platform."
#endif
```

Error in message or in code?

**Comment by thisiskeithb** [comment]
How much proof of life do you need for the SKR 1.3? I can confirm that SD Card (onboard & on LCD), LCD (2004 & 12864), TMC SPI (2130 & 5160), TMC UART (2208/2209), and EEPROM work well on the two SKR 1.3 boards that I have.

**Comment by sky4fly** [comment]
add kinematics delta robot on marlin 2.0 RC1
http://www.cim.mcgill.ca/~paul/clavdelt.pdf

https://www.youtube.com/watch?v=v46Gh64aN_o
https://www.youtube.com/watch?v=1aBubz6Mk2M


smoothieware and reprap firmware supports these kinematics
waiting for washed down on marlin: (

**Comment by pinchies** [comment]
Just want to add the jgaurora A5S and A1, I’ve been debugging marlin 2.0 with both and most things are working great. Bltouch not working yet since servos not working. SD card reliability I would describe as 99%, but that could be due to wire shielding issues. EEPROM on SD working fine. This is all on ZET6 like robin and alfawise. LCD touch working via same method as for those boards too, just a different calibration for screen touch values.

**Comment by RiRilot** [comment]
I'm testing 2.0 on an SKR V1.3 with TMC5160s on all axis and BLTouch. It's all working well apart from the wierd layer shift issue. Happy to test any fixes you might have for that. This is my deveopment printer so can be up and down as needed.

**Comment by coreyfaure** [comment]
> I'm testing 2.0 on an SKR V1.3 with TMC5160s on all axis and BLTouch. It's all working well apart from the wierd layer shift issue. Happy to test any fixes you might have for that. This is my deveopment printer so can be up and down as needed.

Have you been able to get Linear Advance working on your end, or is that something you've tested at all? I can't seem to find a whole lot of results for other tests with Linear Advance on the SKR 1.3. I'm using TMC2208s on the X, Y and Z axis and an A4988 on the E and the calibration patterns won't produce anything usable.

**Comment by RiRilot** [comment]
It's next on my list to test actually. Might have some data tomorrow.

**Comment by RiRilot** [comment]
Linear advance working fine on SKR V1.3. K factor came out at 0.8 with my Bowden setup after calibration.

**Comment by coreyfaure** [comment]
> Linear advance working fine on SKR V1.3. K factor came out at 0.8 with my Bowden setup after calibration.

Which drivers do you have?

**Comment by RiRilot** [comment]
TMC5160s in SPI mode on X, Y, and Z. A4988 on E0.

**Comment by thordarsen** [comment]
On my SKR1.3 Linear Advance works fine with my Titan Aero setup. (2208 XY, A4988 ZE; 0.13 PLA, 0.18 PETG, 0.26 TPU) I've not had much luck when using a Bowden setup, but never tried anything close to 0.8 either.

On another topic, is there a good way to figure out the pins for STM32F1 (03RCT6)? I got an XVico X1 dirt cheap mostly for parts. The main board is discussed here: https://tinkerfiddle.blogspot.com/2019/
I've dug a little further - I know its running DLion firmware - V4 or later. And its based on the DLion Mini board ( https://item.taobao.com/item.htm?spm=a1z10.5-c-s.w4004-17760143202.4.256738f4qt6iZ8&id=592299690248 ) which is I think made by logeek.cn. They make the DLion source available (V04) but the board schematic and pins.h are for the non mini variant. And my extensive use Google Translate has run into a wall getting the info for my board. (I can post the pins.h file and schematics for the non Mini board if anyone wants it)

**Comment by coreyfaure** [comment]
> On my SKR1.3 Linear Advance works fine with my Titan Aero setup. (2208 XY, A4988 ZE; 0.13 PLA, 0.18 PETG, 0.26 TPU) I've not had much luck when using a Bowden setup, but never tried anything close to 0.8 either.

Could you share your config files for your SKR 1.3 and 2208 setup? I've been wrestling with this issue for three weeks and I'm very curious as to what's setup wrong on my end.

**Comment by thordarsen** [comment]
> 
> 
> > On my SKR1.3 Linear Advance works fine with my Titan Aero setup. (2208 XY, A4988 ZE; 0.13 PLA, 0.18 PETG, 0.26 TPU) I've not had much luck when using a Bowden setup, but never tried anything close to 0.8 either.
> 
> Could you share your config files for your SKR 1.3 and 2208 setup? I've been wrestling with this issue for three weeks and I'm very curious as to what's setup wrong on my end.

Here you go. My 2208's are currently in standalone mode because I haven't wrangled up a soldering iron yet (why couldn't they just put a jumper/DIP switch there for UART?)

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/3437303/Configuration.zip)

**Comment by coreyfaure** [comment]
> 
> 
> > > On my SKR1.3 Linear Advance works fine with my Titan Aero setup. (2208 XY, A4988 ZE; 0.13 PLA, 0.18 PETG, 0.26 TPU) I've not had much luck when using a Bowden setup, but never tried anything close to 0.8 either.
> > 
> > 
> > Could you share your config files for your SKR 1.3 and 2208 setup? I've been wrestling with this issue for three weeks and I'm very curious as to what's setup wrong on my end.
> 
> Here you go. My 2208's are currently in standalone mode because I haven't wrangled up a soldering iron yet (why couldn't they just put a jumper/DIP switch there for UART?)
> 
> [Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/3437303/Configuration.zip)

Thank you for this. I duplicated your settings (even down to putting my drivers back in standalone mode) and I'm still getting strange behavior. Do you happen to have a 24V power supply or 12V? I've got a 24V and I can't help but wonder if something is wrong regarding that.

**Comment by thordarsen** [comment]
> Thank you for this. I duplicated your settings (even down to putting my drivers back in standalone mode) and I'm still getting strange behavior. Do you happen to have a 24V power supply or 12V? I've got a 24V and I can't help but wonder if something is wrong regarding that.

24V - Ender 3 (originally)

**Comment by thinkyhead** [comment]
Grand Central support has arrived with #14749

**Comment by simon-jouet** [comment]
### Tests

Okay now that I have a new ESP32 board, I went through the list to check what is working and what's not. All of these tests have been done against fec52e61ea22640d70d7aaa0c8c235958cfe4b24

- **SDCard**: working fine managed to complete a print without problems.
- **LCD**: 
  - **SSD1306** OLED over I2C is working. thanks @Idorobots 
  - **HD44780** from a RepRap Smart Discount controller is working fine, the rotary encoder as well. But I wouldn't really recommend this as it uses a *lot* of pins which are quite precious on the ESP32. For this to work I had to remove LiquidCrystal from the platformio `lib_ignore`. I think this should be removed by default as it's not causing issues.
  - I might try an ILI9341 over SPI at some point (#14578)
- **TMC SPI**: Working fine with TMC2130, currently running TMC2130 in X, Y, TMC2130_standalone in Z and a DRV8825 in E. The SPI bus is shared with the SDCard.
- **TMC UART**: Cannot test, I was planning to order TMC 2209 but looks like most shops are out of stock
- **EEPROM**: It does work but we should move the current implementation from SPIFFS to NVS
- **WiFi**:
  - **OTA**: is working fine (and it's very very nice), this should probably be improved to not allow firmware upload while printing.
  - **WebUI**: is not great, it kinds of work, I started the SDprint from that BUT some messages do crash the socket (e.g M115). The current UI doesn't auto reconnect to the socket. Anyway, I need to get back to @luc-github about this to see how we can go forward (probably replace current code).
- **MAX6675**: I can't test
- **NeoPixel**: Working fine
- **FAST_PWM_FAN**: The ESP32 uses ledc hardware PWM already so this should work fine, we just need to implement the set_pwm_frequency and set_pwm_duty in the HAL

### Example OTA config for platformIO
```
upload_protocol = espota
upload_port = 10.66.0.102
upload_flags = -p 3232
```

### SSD1306 deps changes

I should create a MR for this but in the meantime

```diff
 lib_deps =
+  ${common.lib_deps}
   https://github.com/me-no-dev/AsyncTCP.git
   https://github.com/me-no-dev/ESPAsyncWebServer.git
 lib_ignore  =
-  LiquidCrystal_I2C
   LiquidCrystal
-  NewliquidCrystal
-  LiquidTWI2
   TMC26XStepper
   c1921b4
-  SailfishLCD
-  SailfishRGB_LED

```

### RepRap Discount Smart Controller
![IMG_4027](https://user-images.githubusercontent.com/832495/62006076-de262100-b133-11e9-9519-d8efff8e95a4.JPG)

### NeoPixel (ws2812b with a dodgy setup)
![IMG_4030](https://user-images.githubusercontent.com/832495/62006078-eed69700-b133-11e9-9a57-8ffe1e876f1e.JPG)

**Comment by GMagician** [comment]
Only things that really are tested in AGCM4 and that seems to work are: onboard SD and onboard eeprom
LCD is not working yet (missing u8 libs code), servo have some issue with original library (doesn't work, doesn't compile but there are some PR that should fix these) while tmc may work but are not tested because I don't have any shield and any stepstick to test.

**Comment by Idorobots** [comment]
> SSD1306 OLED over I2C is not working because u8glib doesn't support the ESP32. Using u8g2 would solve this but it's not currently used in Marlin

I did implement that support in the Marlin fork of u8glib here: https://github.com/MarlinFirmware/U8glib-HAL/pull/7 Unlike on LPC (see https://github.com/MarlinFirmware/Marlin/issues/13550) on ESP32 it does seem to be stable.

**Comment by simon-jouet** [comment]
> > SSD1306 OLED over I2C is not working because u8glib doesn't support the ESP32. Using u8g2 would solve this but it's not currently used in Marlin
> 
> I did implement that support in the Marlin fork of u8glib here: [MarlinFirmware/U8glib-HAL#7](https://github.com/MarlinFirmware/U8glib-HAL/pull/7) Unlike on LPC (see #13550) on ESP32 it does seem to be stable.

~~Thanks I will check that at some point, I managed to compile fine when I did the tests but I had a crash loop when Marlin tried to initialise the screen. I assumed it was because of u8glib compatibility but apparently not. The crash was in u8gInit if i remember correctly.~~

EDIT: working fine with @Idorobots changes

**Comment by 0x3333** [comment]
What do you guys means with:

>  STM32: Needs software serial

I'm working in a Software UART to a side project, for Chibios OS on a STM32F1(Probably works on all STM32), and it may fit, depending on your needs. It uses a Timer that works 3 times the baud frequency, all interrupt code, plain simple but highly effective. Probably will finish until end of month.

I'm not familiar with Marlin, so if my question is dumb, just let me know.

**Comment by tpruvot** [comment]
Serial (and multiple serials) seems to works properly on the F1. What "could" be missing is the direct USB pins, printers seems to generally use a dedicated usb/serial chip via a serial channel. But i say could because pins are not linked/used on my boards.

**Comment by 0x3333** [comment]
Well, if that's the case, it is not a "Software serial" that is needed... odd.

**Comment by pinchies** [comment]
@thinkyhead - core function things (LCD, USB serial, SD, Touch, SD-on-eeprom) are working great now on A5S/A1. Can we add them to the supported list to RC1?

**Comment by gloomyandy** [comment]
Software serial is useful for talking to devices like the TMC2208 etc drivers (less so with the TMC2209 which allows multiple drivers to share a UART), when you may need five or more serial interfaces (to talk to the drivers) plus perhaps one or more to talk to say a TFT screen and perhaps another to talk to Octoprint or whatever. So you may need 7 UART devices. I'm not sure how many UARTS the STM devices support, or how easy it is to arrange for the pins to be available (they may be shared with other devices that are also needed like spi/i2c/adc etc.). On the LPC176x chips there simply are not enough hardware UARTS available and so we use software serial (@0x3333 you may want to take a look at how it is implemented for the LPC176x, sounds similar to how you are doing things).

**Comment by 0x3333** [comment]
@gloomyandy yeah, I thought that was the reason. The LPC176X is using arduino software serial, and it is blocking, isn't it?

**Comment by gloomyandy** [comment]
@0x3333 No it is written for the LPC176x It uses the same API as the Arduino (so that it can be used with things like the TMC library). Everything is driven from a timer interrupt. It is sort of blocking on output, but only because the output buffer is only a single byte (like the Arduino version), but it only blocks if the previous character is still being sent. The source is here: https://github.com/p3p/pio-framework-arduino-lpc176x/blob/master/cores/arduino/SoftwareSerial.cpp

Ignore the header comment, that is from the original Arduino software serial code and I didn't get around to updating it before the code was merged.

**Comment by GMagician** [comment]
Please, to whomever can change this post, grandcentral M4 should have ⚠️ on TMC and lcd and 🛑 on WiFi and Max6675

**Comment by kresike** [comment]
> > Does the coming soon BigtreeTech SKR Pro is so much too far from the current Marlin status ?
> > I still don't have the "big picture" of Marlin V2, so sorry in advance if it's a silly question. In fact, I even don't know if STM32F407ZGT6 already exists in Marlin ecosystem.
> 
> It exists. I have a board on the way. If its just a matter of a pin file, config samples ect and the HAL is good, it just might make it in. Time will tell.

I have a printer running on a BIGTREETECH SKR PRO with TMC5160's (X,Y) and TMC2130's (Z1,Z2,E) , everything seems ok except for linear advance, it doesn't extrude properly with LA enabled. Still trying to work that out.

**Comment by thinkyhead** [comment]
> doesn't extrude properly with LA enabled

Be sure to disable spreadCycle on E as this seems to play havoc with LA.

**Comment by gloomyandy** [comment]
@thinkyhead are you sure you mean disable spreadcycle? From what I've seen it is stealthchop that causes problems and spreadcyle usually works.

**Comment by pinchies** [comment]
@thinkyhead - core function things (LCD, USB serial, SD, Touch, SD-on-eeprom) have been working great now on JGAurora A5S/A1 for a while. Can we add them to the supported list to RC1? Different board, but same chip and status as Robin 2.4.

**Comment by jjameshodgson** [comment]
Are there any plans to include support for the 12bit ADCs on these boards? I've tried to switch from Smoothieware to Marlin on my MKS SBASE but the temps are nowhere near as smooth and none of the settings in Marlin seem to make it as stable as SW.

**Comment by jediminer543** [comment]
With respect to the FAST_PWM_FAN; the HAL doesn't support the DUE yet; but does support the LPC176x; as the DUE HAL lacks the `void set_pwm_frequency(const pin_t pin, int f_desired)` definition anywhere in it's HAL. (and similarly the LPC176x does)

**Comment by RiRilot** [comment]
Just thought I'd chip in an say that I've been running 2.0 on an SKR V1.3 for about 2 months now. Configured with linear advance and TMC5160s in SPI mode. No issues at all. It's solid and believe me, I hammer it daily.

**Comment by dan-and** [comment]
> STM32: Needs software serial

Just an overview of the Status: 
stm32duino added that feature on the 11. September 2019 and released it with stm32duino v1.7.0 . 
platform-ststm32 merged stm32duino v1.7.0 today (1. October 2019), so when the the next release of ststm32 (something like 5.6.1 or 5.7.0) will include the software serial for the STM32. 

So things are moving forward :-)

**Comment by ghost** [comment]
I2C GPIO expanders for Teensy 3.5 would be great. The chip is pretty powerful and could enable shields with more features.
PCF8574 is pretty simple and common.

**Comment by extesy** [comment]
I guess this issue can be closed due to https://github.com/MarlinFirmware/Marlin/releases/tag/2.0.0

**Comment by boelle** [comment]
@thinkyhead do you agree with @extesy ^^^

**Comment by boelle** [comment]
will close this one as 2.0 is out

**Comment by chepo92** [comment]
I see Duet 2 Wifi + X5  (SAM4E8E) marked as in alpha, is there a HAL for it? Anyone working on it ?

**Comment by chepo92** [comment]
@thinkyhead can you copy and send me your original message (I want it with the Markdonw style), for documentation purposes

**Comment by Hedda** [comment]
> @thinkyhead can you copy and send me your original message (I want it with the Markdonw style), for documentation purposes

FYI, you can click the three dots (...) in the top-right of a message and choose "quote reply" to copy it.

**Comment by chepo92** [comment]
> 
> 
> > @thinkyhead can you copy and send me your original message (I want it with the Markdonw style), for documentation purposes
> 
> FYI, you can click the three dots (...) in the top-right of a message and choose "quote reply" to copy it.

---

I had the same thought but I only get the source Markdown of the list, not tables: 

---
> 
> 
> Let's coordinate the Marlin 2.0 RC1 release and track remaining work for this first release candidate.
> ### AVR
> Board 	MCU 	SD Card 	LCD 	TMC SPI 	TMC UART 	EEPROM 	WiFi 	Max6675 	Neopixel
> [RAMPS](https://reprap.org/wiki/RAMPS_1.4), [RAMBo](https://reprap.org/wiki/Rambo) etc. 	AVR 	green_heart 	green_heart 	green_heart 	green_heart 	green_heart 	green_heart 	green_heart 	green_heart
> ### SAM
> Board 	MCU 	SD Card 	LCD 	TMC SPI 	TMC UART 	EEPROM 	WiFi 	Max6675 	Neopixel

>       
>       * [x]  AVR
>       * [x]  DUE
>       * [ ]  LPC176x
>       * [ ]  ESP32
>       * [ ]  STM32
>       * [ ]  STM32F1
>       * [ ]  STM32F4
>       * [ ]  STM32F7 - one or two boards
>       * [ ]  Teensy 3.1/3.2
>       * [ ]  Teensy 3.5/3.6
